xrtl/ui/imgui_overlay.cc

```diff
@@ -0,0 +1,684 @@
+            static_cast<int>(draw_cmd.ClipRect.w - draw_cmd.ClipRect.y)};
+        if (scissor_rect != previous_scissor_rect) {
+          rpe->SetScissor(scissor_rect);
+          previous_scissor_rect = scissor_rect;
+        }
+
+        // Bind the texture resource used (if any).
+        if (draw_cmd.TextureId) {
+          // TODO(benvanik): cache resources (especially for font).
+          // TODO(benvanik): combined image sampler resource to simplify?
+          ref_ptr<gfx::ImageView> image_view{
+              reinterpret_cast<gfx::ImageView*>(draw_cmd.TextureId)};
+          std::vector<ResourceSet::BindingValue> binding_values;
+          binding_values.push_back(
+              {image_view, Image::Layout::kGeneral, nearest_sampler_});
+          auto resource_set = context_->CreateResourceSet(
+              render_pipeline_layout_, binding_values);
+          if (!resource_set) {
+            LOG(ERROR) << "Unable to create resource set";
+            return;
+          }
+          rpe->BindResourceSet(resource_set);
+        } else {
+          // TODO(benvanik): a default set with a white pixel texture?
+          rpe->BindResourceSet(nullptr);
+        }
+
+        // Issue draw for the buffer range.
+        rpe->DrawIndexed(draw_cmd.ElemCount, 1, index_offset, 0, 0);
+      }
+      index_offset += draw_cmd.ElemCount;
+    }
+  }
+
+  command_buffer_->EndRenderPass(std::move(rpe));
+
+  // Flush and unmap the buffers before we use them.
+  vertex_mapping.reset();
+  index_mapping.reset();
+}
+
+ref_ptr<gfx::QueueFence> ImGuiOverlay::EndFrame(
+    ref_ptr<gfx::QueueFence> wait_fence) {
+  DCHECK_EQ(ImGui::GetCurrentContext(), imgui_context_);
+
+  // Prepare the command buffer that RenderDrawLists will write into.
+  command_buffer_ = context_->CreateCommandBuffer();
+
+  // Flush all imgui commands. This will call our RenderDrawLists fn.
+  ImGui::Render();
+
+  // TODO(benvanik): get outputs and redirect to control/etc.
+  // io->WantCaptureMouse
+  // io->WantCaptureKeyboard
+  // io->WantTextInput
+
+  // Clear the current context until the next frame.
+  framebuffer_.reset();
+  ImGui::SetCurrentContext(nullptr);
+
+  // Submit the command buffer.
+  ref_ptr<gfx::QueueFence> signal_fence = context_->CreateQueueFence();
+  auto submit_result = context_->Submit(
+      std::move(wait_fence), std::move(command_buffer_), signal_fence);
+  command_buffer_.reset();
+  switch (submit_result) {
+    case gfx::Context::SubmitResult::kSuccess:
+      break;
+    case gfx::Context::SubmitResult::kCommandBufferReused:
+      LOG(ERROR)
+          << "Unable to submit imgui render commands: command buffer reused";
+      return nullptr;
+    case gfx::Context::SubmitResult::kDeviceLost:
+      LOG(ERROR) << "Unable to submit imgui render commands: device lost";
+      return nullptr;
+  }
+
+  return signal_fence;
+}
+
+namespace {
+
+void PopulateEventData(const KeyboardEvent& ev, bool is_down, ImGuiIO* io) {
```

Comment:
They're private implementation functions and don't need any state from ImGuiOverlay so there's no reason to expose them in the header.